### PR TITLE
Copy boundary info from other meshes

### LINKED ISSUE #15185: MeshCollectionGenerator should copy side sets etc.
## Reason
MeshCollectionGenerator does not copy the side set assignments (boundary info) to the combined mesh.

## Design
Add code to copy boundary info. 

## Impact
Get up to expected capability.

### PR DESCRIPTION
**Pull request by dschwen**
This makes sure boundary IDs and names are copied over when combining meshes. The generator will now throw an error if different names for the same ID exist on the different input meshes.

To make things easier I also added a capability to GeneratedMeshGenerator to permit prefixing the autogenerated sideset names (top, bottom, left, right, etc.) and to add an offset to the sideset IDs. 

With these two changes I was able to shorten a bunch of contact tests. Ping @YaqiWang , who suggested something similar a while ago.

Closes #15185

### CODE REVIEW

**Comment by dschwen** [comment]
Had to regold the test and made it `-pedantic`. That ensures the number of nodes in the sidesets is the same. We should probably add that option to a few more meshgenerator tests.

File: modules/contact/test/tests/mortar_tm/2d/ad_frictionless_fir/finite.i [hunk 1]
```diff
@@ -17,16 +17,11 @@ name = 'finite'
     nx = 2
     ny = 67
     elem_type = ${elem}
-  [../]
-  [./plank_sidesets]
-    type = RenameBoundaryGenerator
-    input = plank
-    old_boundary_id = '0 1 2 3'
-    new_boundary_name = 'plank_bottom plank_right plank_top plank_left'
+    sideset_name_prefix = plank
```
Comment (bwspenc) [inline modules/contact/test/tests/mortar_tm/2d/ad_frictionless_fir/finite.i:20]: Much better!

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: framework/src/meshgenerators/GeneratedMeshGenerator.C [hunk 2]
```diff
@@ -97,7 +103,11 @@ GeneratedMeshGenerator::GeneratedMeshGenerator(const InputParameters & parameter
     _gauss_lobatto_grid(getParam<bool>("gauss_lobatto_grid")),
     _bias_x(getParam<Real>("bias_x")),
     _bias_y(getParam<Real>("bias_y")),
-    _bias_z(getParam<Real>("bias_z"))
+    _bias_z(getParam<Real>("bias_z")),
+    _sideset_name_prefix(isParamValid("sideset_name_prefix")
```
[the remark this answers is not in the document]
Comment (dschwen) [reply framework/src/meshgenerators/GeneratedMeshGenerator.C:107]: Yeah, not a problem.

File: framework/src/meshgenerators/GeneratedMeshGenerator.C [hunk 3]
```diff
@@ -176,6 +186,19 @@ GeneratedMeshGenerator::generate()
...
   }
 
+  // rename and shift boundaries
+  BoundaryInfo & boundary_info = mesh->get_boundary_info();
+  const auto & mesh_boundary_ids = boundary_info.get_boundary_ids();
+  for (auto rit = mesh_boundary_ids.rbegin(); rit != mesh_boundary_ids.rend(); ++rit)
+  {
+    boundary_info.sideset_name(*rit + _sideset_id_offset) =
+        _sideset_name_prefix + boundary_info.sideset_name(*rit);
+    boundary_info.nodeset_name(*rit + _sideset_id_offset) =
```
[the remark this answers is not in the document]
Comment (dschwen) [reply framework/src/meshgenerators/GeneratedMeshGenerator.C:196]: Yeah, I'm really not clear when to use boundary vs. sideset vs. nodeset. The existing code does not help me understand this either.

File: framework/src/meshgenerators/GeneratedMeshGenerator.C [hunk 4]
```diff
@@ -71,6 +72,11 @@ GeneratedMeshGenerator::validParams()
       "bias_z>=0.5 & bias_z<=2",
       "The amount by which to grow (or shrink) the cells in the z-direction.");
 
+  params.addParam<std::string>("sideset_name_prefix",
+                               "If provided, prefix the built in sitdeset names with this string");
+  params.addParam<boundary_id_type>(
+      "sideset_id_offset", 0, "This offset is added to the generated boundary IDs");
```
[the remark this answers is not in the document]
Comment (dschwen) [reply framework/src/meshgenerators/GeneratedMeshGenerator.C:78]: No, it's actually the opposite. Both sidesets and nodesets are prefixed. So this should probably be called `boundary` everywhere. The IDs are - I believe - the same for node and sidesets (`MeshTools::Modification::change_boundary_id` changes the ID on both the nodes and the sides).

File: framework/src/meshgenerators/GeneratedMeshGenerator.C (hunk 2 above)
[the remark this answers is not in the document]
Comment (dschwen) [reply framework/src/meshgenerators/GeneratedMeshGenerator.C:107]: The no default scheme follows our other prefixes and allows us to more cleanly enforce a separating `_` between prefix and name,

File: framework/src/meshgenerators/GeneratedMeshGenerator.C (hunk 3 above)
[the remark this answers is not in the document]
Comment (dschwen) [reply framework/src/meshgenerators/GeneratedMeshGenerator.C:196]: I guess boundary the the more general term and nodeset and sideset are the two variants of specified boundaries.

File: framework/include/meshgenerators/GeneratedMeshGenerator.h [hunk 5]
```diff
@@ -40,17 +40,27 @@ class GeneratedMeshGenerator : public MeshGenerator
...
+  /**
+   * The amount by which to bias the cells in the x,y,z directions.
+   * Must be in the range 0.5 <= _bias_x <= 2.0.
+   * _bias_x < 1 implies cells are shrinking in the x-direction.
+   * _bias_x==1 implies no bias (original mesh unchanged).
+   * _bias_x > 1 implies cells are growing in the x-direction.
+   */
+  const Real _bias_x, _bias_y, _bias_z;
+
+  /// prefix string for the side set names
```
Comment (bwspenc) [inline framework/include/meshgenerators/GeneratedMeshGenerator.h:61]: All comments are lies!

side set -> boundary

Am I being too picky?

**Review by bwspenc — DISMISSED** [review]
Other than my very pedantic comment, looks good.

**Comment by bwspenc** [comment]
@dschwen I think this was pretty much ready to go. Just fix those couple of minor things and resubmit, and I'll merge it.

**Comment by dschwen** [comment]
Argh, I thought I was losing my mind! I thought it had already been merged (I need that functionality now).

**Review by bwspenc — APPROVED** [review]

**Comment by lindsayad** [comment]
@dschwen could you take a look at https://civet.inl.gov/job/559628/ ?

**Pull request merged**